From 009fce98be95c1674a697b850157c59bacda3168 Mon Sep 17 00:00:00 2001
From: Eric Blake <eblake@redhat.com>
Date: Mon, 21 Feb 2011 15:05:24 -0700
Subject: [PATCH] security: avoid memory leak

Leak introduced in commit d6623003.

* src/qemu/qemu_driver.c (qemuSecurityInit): Avoid leak on failure.
* src/security/security_stack.c (virSecurityStackClose): Avoid
leaking component drivers.
Index: libvirt-0.8.8/src/qemu/qemu_driver.c
===================================================================
--- libvirt-0.8.8.orig/src/qemu/qemu_driver.c	2011-03-15 11:50:54.000000000 -0500
+++ libvirt-0.8.8/src/qemu/qemu_driver.c	2011-03-15 11:51:01.000000000 -0500
@@ -1017,8 +1017,11 @@
             goto error;
 
         if (!(driver->securityManager = virSecurityManagerNewStack(mgr,
-                                                                   dac)))
+                                                                   dac))) {
+
+            virSecurityManagerFree(dac);
             goto error;
+        }
     } else {
         driver->securityManager = mgr;
     }
Index: libvirt-0.8.8/src/security/security_stack.c
===================================================================
--- libvirt-0.8.8.orig/src/security/security_stack.c	2011-03-15 11:50:54.000000000 -0500
+++ libvirt-0.8.8/src/security/security_stack.c	2011-03-15 11:51:01.000000000 -0500
@@ -61,8 +61,13 @@
 }
 
 static int
-virSecurityStackClose(virSecurityManagerPtr mgr ATTRIBUTE_UNUSED)
+virSecurityStackClose(virSecurityManagerPtr mgr)
 {
+    virSecurityStackDataPtr priv = virSecurityManagerGetPrivateData(mgr);
+
+    virSecurityManagerFree(priv->primary);
+    virSecurityManagerFree(priv->secondary);
+
     return 0;
 }
 
